src/libraries/System.Linq/src/System/Linq/MaxMin.cs

```diff
@@ -0,0 +1,118 @@
+        {
+            T value;
+
+            if (source.TryGetSpan(out ReadOnlySpan<T> span))
+            {
+                if (span.IsEmpty)
+                {
+                    ThrowHelper.ThrowNoElementsException();
+                }
+
+                if (!Vector128.IsHardwareAccelerated || span.Length < Vector128<T>.Count)
+                {
+                    value = span[0];
+                    for (int i = 1; i < span.Length; i++)
+                    {
+                        if (TMinMax.Compare(span[i], value))
+                        {
+                            value = span[i];
+                        }
+                    }
+                }
+                else if (!Vector256.IsHardwareAccelerated || span.Length < Vector256<T>.Count)
+                {
+                    ref T current = ref MemoryMarshal.GetReference(span);
+                    ref T lastVectorStart = ref Unsafe.Add(ref current, span.Length - Vector128<T>.Count);
+
+                    Vector128<T> best = Vector128.LoadUnsafe(ref current);
+                    current = ref Unsafe.Add(ref current, Vector128<T>.Count);
+
+                    while (Unsafe.IsAddressLessThan(ref current, ref lastVectorStart))
+                    {
+                        best = TMinMax.Compare(best, Vector128.LoadUnsafe(ref current));
+                        current = ref Unsafe.Add(ref current, Vector128<T>.Count);
+                    }
+                    best = TMinMax.Compare(best, Vector128.LoadUnsafe(ref lastVectorStart));
+
+                    value = best[0];
+                    for (int i = 1; i < Vector128<T>.Count; i++)
+                    {
+                        if (TMinMax.Compare(best[i], value))
+                        {
+                            value = best[i];
+                        }
+                    }
+                }
+                else
+                {
+                    ref T current = ref MemoryMarshal.GetReference(span);
+                    ref T lastVectorStart = ref Unsafe.Add(ref current, span.Length - Vector256<T>.Count);
+
+                    Vector256<T> best = Vector256.LoadUnsafe(ref current);
+                    current = ref Unsafe.Add(ref current, Vector256<T>.Count);
+
+                    while (Unsafe.IsAddressLessThan(ref current, ref lastVectorStart))
+                    {
+                        best = TMinMax.Compare(best, Vector256.LoadUnsafe(ref current));
+                        current = ref Unsafe.Add(ref current, Vector256<T>.Count);
+                    }
+                    best = TMinMax.Compare(best, Vector256.LoadUnsafe(ref lastVectorStart));
+
+                    value = best[0];
+                    for (int i = 1; i < Vector256<T>.Count; i++)
+                    {
+                        if (TMinMax.Compare(best[i], value))
+                        {
+                            value = best[i];
+                        }
+                    }
```

Comment:
I'm wondering if this code could be somehow generalized as well since it's effectively identical as Vector128 (or what could we change to make it generalizable if it currently is)

Reply:
At the moment, no. Tanner and I spoke about whether it would make sense to add an IVector interface. I still need to open an issue for that, but it doesn't help right now.